core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java

```diff
@@ -77,22 +78,22 @@ public long startRecovery(IndexShard indexShard, DiscoveryNode sourceNode,
      */
     public void resetRecovery(long id, ShardId shardId) throws IOException {
         try (RecoveryRef ref = getRecoverySafe(id, shardId)) {
-            final RecoveryTarget copy = new RecoveryTarget(ref.status);
             // instead of adding complicated state to RecoveryTarget we just flip the
             // target instance when we reset a recovery, that way we have only one cleanup
             // path on the RecoveryTarget and are always within the bounds of ref-counting
             // which is important since we verify files are on disk etc. after we have written them etc.
+            RecoveryTarget status = ref.status();
+            RecoveryTarget resetRecovery = status.resetRecovery();
             boolean success = false;
             try {
-                if (onGoingRecoveries.replace(ref.status.recoveryId(), ref.status, copy) == false) {
-                    throw new IllegalStateException("failed to reset recovery");
+                if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
+                    throw new IllegalStateException("failed to replace recovery target");
                 }
-                ref.status.close();
-                copy.indexShard().performRecoveryRestart();
                 success = true;
             } finally {
                 if (success == false) {
-                    copy.cancel("reset failed");
+                    onGoingRecoveries.remove(id);
```

Comment:
nit: now that we folded the close and performRecoveryRestart into `status.resetRecovery()` we don't need the success pattern anymore. This can be:

```
if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
       resetRecovery.cancel("replace failed");
       throw new IllegalStateException("failed to replace recovery target");
}

```